forgerock-auth/src/main/java/org/forgerock/android/auth/WebAuthnDataRepository.kt

```diff
@@ -0,0 +1,151 @@
+        return EncryptedPreferences.getInstance(context, ORG_FORGEROCK_V_2_WEBAUTHN)
+    }
+
+    private fun migrate(sharedPreferences: SharedPreferences): SharedPreferences {
+        val encryptedPreferences = getNewSharedPreferences()
+        sharedPreferences.getString(ALLOW_CREDENTIALS, null)?.apply {
+            val map = mutableMapOf<String, MutableSet<String>>()
+            val array = JSONArray(this)
+            for (o in array.iterator<JSONObject>()) {
+                fromJson(o).apply {
+                    (map[this.rpid] ?: mutableSetOf<String>().also {
+                        map[this.rpid] = it
+                    }).add(this.toJson().put("created", System.currentTimeMillis()).toString())
+                }
+            }
+            encryptedPreferences.edit().apply {
+                map.forEach {
+                    this.putStringSet(it.key, it.value)
+                }
+            }.apply()
+
+            sharedPreferences.edit().remove(ALLOW_CREDENTIALS).apply()
+        }
+        return encryptedPreferences
+    }
+
+    /**
+     * Persist the [PublicKeyCredentialSource], it overrides the existing
+     * [PublicKeyCredentialSource] with the same [PublicKeyCredentialSource.getUserHandle]
+     * * @param source The [PublicKeyCredentialSource] to persist.
+     */
+    fun persist(source: PublicKeyCredentialSource) {
+        if (!this::dataRepository.isInitialized) {
+            warn(TAG, "UsernameLess cannot be supported. No credential will be stored")
+            return
+        }
+        dataRepository.getStringSet(source.rpid, emptySet())!!.apply {
+            val result = mutableSetOf<String>()
+            result.addAll(this)
+            result.add(source.toJson().put("created", System.currentTimeMillis()).toString())
+            dataRepository.edit().putStringSet(source.rpid, result).apply()
+        }
+    }
+
+    /**
+     * Retrieve all [PublicKeyCredentialSource] with the same rpId [PublicKeyCredentialSource.getRpid]
+     *
+     * @param rpId The Relying Party Id
+     * @return All [PublicKeyCredentialSource] with the same rpId [PublicKeyCredentialSource.getRpid]
+     */
+    fun getPublicKeyCredentialSource(rpId: String): List<PublicKeyCredentialSource> {
+        if (!this::dataRepository.isInitialized) {
+            warn(TAG, "UsernameLess cannot be supported. No credential will be stored")
+            return emptyList()
+        }
+        return dataRepository.getStringSet(rpId, emptySet())?.mapNotNull {
+            JSONObject(it)
+        }?.sortedWith(compareByDescending {
+            it.optLong("created", 0)
+        })?.map {
+            fromJson(it)
+        }?: emptyList()
+    }
+
+    fun deleteByUserIdAndRpId(userId: ByteArray, rpId: String) {
+        if (!this::dataRepository.isInitialized) {
+            warn(TAG, "UsernameLess cannot be supported. No credential will be stored")
+            return
+        }
+
```

Comment:
this code can be refactored 

        val storedCredentials = dataRepository.getStringSet(rpId, emptySet())?.filter { !fromJson(JSONObject(it)).userHandle.contentEquals(userId) }?.toSet()
        dataRepository.edit().putStringSet(rpId, storedCredentials).apply()

Reply:
getStringSet can return nullable , so we dont want to crash

Reply:
A default value can be provided during getStringSet, I created an emptySet if not found, but since we do null check for the object, we can return null instead of emptyset.

For the second statement, the storedCredentials may be empty after delete, we have to update the set even it is empty.